### PR TITLE
Asset flow verification in `driver`

### PR DESCRIPTION
**Pull request by sistemd**
Half of https://github.com/cowprotocol/services/issues/1009.

Implements points 1. and 2. of the issue above, in particular:
- Ensure that solutions simulate without reverting.
- Ensure that the solutions don't take any money out of the settlement contract.

I will implement point 3. in a follow-up PR.

### Test Plan

Added automated tests. The existing tests continue to pass.

### CODE REVIEW

File: crates/driver/src/domain/competition/solution/mod.rs [hunk 1]
```diff
@@ -107,16 +107,28 @@ impl Solution {
...
+    /// [`settlement::Verified`].
+    pub async fn verify(
+        &self,
+        eth: &Ethereum,
+        simulator: &Simulator,
+        auction: &competition::Auction,
+    ) -> Result<settlement::Verified, Error> {
+        self.verify_asset_flow()?;
+        self.verify_internalization(auction)?;
+        self.simulate(eth, simulator, auction).await
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/mod.rs:120]: According to the [issue](https://github.com/cowprotocol/services/issues/1009#issuecomment-1387621803) both the simulations for the internalized as well as the uninternalized version need to pass. It's fine to do it in a separate PR; just wanted to point it out so we don't forget about it.

File: crates/driver/src/domain/competition/solution/mod.rs [hunk 2]
```diff
@@ -158,12 +170,65 @@ impl Solution {
...
             gas,
         })
     }
+
+    /// Check that the sum of tokens entering the settlement is not less than
+    /// the sum of tokens exiting the settlement.
+    fn verify_asset_flow(&self) -> Result<(), VerificationError> {
+        let mut sum_inputs = eth::U256::zero();
+        let mut sum_outputs = eth::U256::zero();
+        for interaction in self.interactions.iter() {
```
Comment (nlordell) [inline crates/driver/src/domain/competition/solution/mod.rs:185]: Orders also account for asset flow that we need to verify here, i.e.

```rust
sum_inputs[order.sell.token] += order.sell.amount;
sum_inputs[order.buy.token] -= order.buy.amount;
```

File: crates/driver/src/domain/competition/solution/mod.rs [hunk 3]
```diff
@@ -158,12 +170,65 @@ impl Solution {
...
             access_list,
             gas,
         })
     }
+
+    /// Check that the sum of tokens entering the settlement is not less than
+    /// the sum of tokens exiting the settlement.
+    fn verify_asset_flow(&self) -> Result<(), VerificationError> {
+        let mut sum_inputs = eth::U256::zero();
+        let mut sum_outputs = eth::U256::zero();
```
Comment (nlordell) [inline crates/driver/src/domain/competition/solution/mod.rs:184]: Token conservation should be per token, and not normalized to Ether.

File: crates/driver/src/domain/competition/solution/mod.rs [hunk 4]
```diff
@@ -158,12 +170,65 @@ impl Solution {
...
+        for interaction in self.interactions.iter() {
+            match interaction {
+                Interaction::Custom(interaction) => {
+                    for input in interaction.inputs.iter() {
+                        sum_inputs += self
+                            .reference_amount(input)
+                            .ok_or(VerificationError::AssetFlow)?;
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/driver/src/domain/competition/solution/mod.rs:191]: I don't think we should be normalizing token to Ether with reference prices here. As Martin suggests, this means that all tokens (included liquidity) need reference prices, which isn't the case.

Also, to me, token conservation is a simpler condition of `∀ t ∈ tokens : ∑flux(t) > 0` where `flux(t)` is the flux of tokens into the contract (i.e. +ve value for transfers into the contract, and -ve values for transfers out of the contract).

[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/driver/src/domain/competition/solution/mod.rs:191]: It's only required to return prices for every cow protocol order. If a token is exclusively part of some custom interaction to acquire some other useful tokens the solvers currently wouldn't have to provide a price but would have to with the current token conservation test.

[the remark this answers is not in the document]
Comment (nlordell) [reply crates/driver/src/domain/competition/solution/mod.rs:191]: > reference price coming from the solver

I call the prices that come from the solvers "clearing prices", as they are the actually executed prices on-chain. "reference prices" are the informational prices that come from the autopilot and are always denominated in Ether.

> I was under the impression that the solver needs to return a reference price for every token in its solution.

Only traded tokens (i.e. tokens that appear in user orders). So if you have a user trading A -> C, and interactions for swapping on Uniswap A -> B -> C (so a "1 hop" swap), then only clearing prices for A and C are necessarily part of the solution.

Additionally, clearing price vector does not require normalized prices (i.e. the prices don't need to be denominated in a common unit). For example, imagine a solution with orders:
- USDC -> USDT
- DAI -> FRAX

Noting that all of the tokens are USD stable coins, we can set the prices to be:
- USDC: 1
- USDT: 1
- DAI: 100
- FRAX: 100

I.e. they can be arbitrarily "scaled" if the tokens aren't connected by orders.

**Review by MartinquaXD — APPROVED** [review]
Approving since this will get fixed in a follow up PR.

**Review by nlordell — APPROVED** [review]
